File: web/lib/pricing.js
/**
 * Suggested per-territory prices — the browser counterpart of
 * `core/services/pricing_calculator`.
 *
 * A read-only sanity check, not a push to any store: this never calls App
 * Store Connect or Play Console. The numbers are a starting point to copy in
 * yourself.
 */

import { countryName } from './market.js';

let SPEC = null;

export function loadPricingSpec(specs) {
  SPEC = specs.pricing ?? specs;
}

export function getPricingSpec() {
  if (!SPEC) throw new Error('loadPricingSpec() must be called before suggesting prices');
  return SPEC;
}

export function tierFor(country) {
  const code = country.toLowerCase();
  return getPricingSpec().tiers.find((t) => t.countries.includes(code)) ?? null;
}

/**
 * Every storefront the pricing spec has a tier for, deduplicated. The
 * default set to price when the caller does not name specific countries.
 */
export function knownCountries() {
  return [...new Set(getPricingSpec().tiers.flatMap((t) => t.countries))];
}

export function suggestPrices(basePrice, { model = 'ppp_tier', countries } = {}) {
  const spec = getPricingSpec();
  if (!spec.models[model]) {
    throw new Error(`Unknown pricing model "${model}". Choose from: ${Object.keys(spec.models).join(', ')}.`);
  }

  const codes = (countries ?? knownCountries()).map((c) => c.toLowerCase());
  return codes.map((code) => {
    const tier = tierFor(code);
    const multiplier = tier && model === 'ppp_tier' ? tier.multiplier : 1.0;
    return {
      country: code,
      countryName: countryName(code),
      tierId: tier?.id ?? null,
      tierLabel: tier?.label ?? null,
      multiplier,
      suggestedPrice: Math.round(basePrice * multiplier * 100) / 100,
    };
  });
}

File: src/appmates/cli/commands/pricing.py
"""`appmates pricing` — suggested per-territory prices from one base price."""

from __future__ import annotations

from typing import Annotated

import typer

from appmates.cli.console import ExitCode, console, emit_json, fail
from appmates.core.services.pricing_calculator import suggest_prices
from appmates.core.services.reporting import render_pricing


def pricing(
    base_price: Annotated[
        float, typer.Argument(help="Your price in the base storefront, e.g. 4.99.")
    ],
    base_country: Annotated[
        str, typer.Option("--base-country", help="Storefront the base price is set in.")
    ] = "us",
    model: Annotated[
        str, typer.Option("--model", "-m", help="Pricing model: uniform or ppp_tier.")
    ] = "ppp_tier",
    countries: Annotated[
        str | None,
        typer.Option(
            "--countries", help="Comma-separated storefronts. Defaults to the built-in list."
        ),
    ] = None,
    as_json: Annotated[bool, typer.Option("--json")] = False,
) -> None:
    """Suggest a starting price per storefront from one base price.

    A read-only sanity check, not a push to any store: copy the numbers you
    want into App Store Connect or Play Console yourself.

    [bold]--model uniform[/bold] reprints the same price everywhere, for
    comparison. [bold]--model ppp_tier[/bold] (the default) scales it down in
    lower-income storefronts using coarse World Bank income-group bands —
    an approximation, not exact purchasing-power data.
    """
    codes = [c.strip().lower() for c in countries.split(",") if c.strip()] if countries else None

    try:
        plan = suggest_prices(base_price, base_country=base_country, model=model, countries=codes)
    except ValueError as exc:
        raise fail(str(exc)) from exc

    if as_json:
        emit_json(plan)
    else:
        console.print(render_pricing(plan))

    raise typer.Exit(int(ExitCode.OK))
